akka-stream/src/main/scala/akka/stream/impl/UnfoldResourceSourceAsync.scala

```diff
@@ -107,14 +107,13 @@ import scala.util.control.NonFatal
 
     private def createResource(): Unit = {
       create().onComplete { resource =>
-        createdCallback(resource).failed.foreach {
-          case _: StreamDetachedException =>
-            // stream stopped
-            resource match {
-              case Success(r) =>
-                close(r)
-              case Failure(ex) => throw ex // failed to open but stream is stopped already
-            }
+        createdCallback(resource).failed.foreach { _ =>
+          // stream stopped
+          resource match {
+            case Success(r) =>
+              close(r)
```

Comment:
:+1: to always closing the resource

I would sort of have expected that we'd have to re-throw the exception here to make the stream fail in the "when reading an element throws" - but indeed this is already taken care of in `createdCallback` :+1:

Reply:
Humhum. So, the `invokeWithFeedback` future will fail with `StreamDetached` if the stage stopped before processing the invokation, but it will also fail with the thrown exception if processing the callback throws. So perhaps we should handle those two cases separately and keep the exception in the latter case?

Reply:
Added comment to the wrong review comment, was a reply to the other one.

Reply:
Actually it is tricker than that, I noticed when I handle `StreamDetached` separately.

If the resource is successfully opened, the first element fails and strategy is restart, the restarting calls `close` but does not deal with exceptions like it should (I think) which ends up with that exception reaching here.

Reply:
Ah, with your latest update this is easier to understand from the code :+1: